code401challenges/src/main/java/linkedList/LinkedList.java

```diff
@@ -50,12 +50,82 @@ public String toString() {
         return String.valueOf(listValues);
     }
 
-    //Helper function getHead
-    //getHead
-    //Helper function getHeadValue
-    //getHeadValue
+    // Write the following methods for the Linked List class:
+    // .append(value) which adds a new node with the given value to the end of the list
+    public void append(T value) {
+        Node<T> currentNode = this.head;
+
+        //If there are no nodes in the list
+        if (currentNode == null) {
+            Node newNode = new Node(value);
+            currentNode = newNode;
+        }
+
+        //Traverse to the end of the list and then append
+        while (currentNode.next != null) {
+            currentNode = currentNode.next;
+            if (currentNode.next == null) {
+                Node newNode = new Node(value);
+                currentNode.next = newNode;
+            }
+        }
+    }
+
+    // .insertBefore(value, newVal) which add a new node with the given newValue immediately before the first value node
+    // The boolean will let us know if we found the value in the list or not
+    public boolean insertBefore(T value, T newValue) {
+        Node currentNode = this.head;
+
+        //If the value is found in the first node, make the new node the head and point it at the previous head
+        if (currentNode == null) {
+            return false;
+        } else if (currentNode.value == value) {
+            Node newNode = new Node(newValue, head);
+            head = newNode;
+            return true;
+        }
+
+        //Traverse the list to look for the node with the desired value
+        while (currentNode.next.value != value && currentNode.next != null) {

```

Comment:
you are checking currentNode.next.value before you find if currentNode.next is null. This throws a null pointer exception if next is null and you access its value property

Reply:
Looks like I refactored this one as well.
while (currentNode != null) {
            if (currentNode.value == desiredValue) {
                return true;
            }
            currentNode = currentNode.next;
        }
        return false;